examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py

```diff
@@ -0,0 +1,264 @@
+# Copyright (c) Qualcomm Innovation Center, Inc.
+# All rights reserved
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import gc
+import json
+import os
+from multiprocessing.connection import Client
+
+import executorch.backends.qualcomm.python.PyQnnManagerAdaptor as PyQnnManagerAdaptor
+
+import torch
+from executorch.backends.qualcomm.serialization.qnn_compile_spec_schema import (  # noqa: F401
+    QcomChipset,
+)
+from executorch.backends.qualcomm.utils.utils import (
+    canonicalize_program,
+    from_context_binary,
+    generate_htp_compiler_spec,
+    generate_qnn_executorch_compiler_spec,
+    generate_qnn_executorch_option,
+)
+from executorch.examples.qualcomm.utils import (
+    setup_common_args_and_variables,
+    SimpleADB,
+)
+from executorch.exir.backend.backend_api import to_backend
+from executorch.exir.passes.memory_planning_pass import MemoryPlanningPass
+
+
+def main(args):
+    os.makedirs(args.artifact, exist_ok=True)
+
+    target_names = (
+        [
+            f"llama_v3_8b_chat_quantized_PromptProcessor_{i}_Quantized.bin"
+            for i in range(1, 6)
+        ]
+        if args.use_prompt_processor
+        else [
+            f"llama_v3_8b_chat_quantized_TokenGenerator_{i}_Quantized.bin"
+            for i in range(1, 6)
+        ]
+    )
+
+    # common part for compile & inference
+    backend_options = generate_htp_compiler_spec(
+        use_fp16=False,
+        use_multi_contexts=True,
+    )
+    compiler_specs = generate_qnn_executorch_compiler_spec(
+        soc_model=getattr(QcomChipset, args.model),
+        backend_options=backend_options,
+        is_from_context_binary=True,
+    )
+
+    pte_name = (
+        "qaihub_llama3_8b_prompt"
+        if args.use_prompt_processor
+        else "qaihub_llama3_8b_token"
+    )
+    if args.pre_gen_pte is None:
+        # create custom operators as context loader
+        bundle_programs = [
+            from_context_binary(f"{args.context_binaries}/{target}", f"ctx_loader_{i}")
+            for i, target in enumerate(target_names)
+        ]
+        # lower with QnnBackend
+        lowered_modules = [
+            to_backend("QnnBackend", prog["edge_program"], compiler_specs)
+            for prog in bundle_programs
+        ]
+        # TODO: QNN seems to have an expected spill fill size that can be found through log.
+        # Find a way to set this value instead of manually go through the log to retrieve the value.
+        custom_spill_fill = 128974848 if args.use_prompt_processor else 3932160
+        # setup spill-fill buffer for relieving runtime memory usage
```

Comment:
Do we know the ram usage and latency number?

Reply:
Hi @cccclai ,
Thanks a lot for reviewing the PR.

- Latency: We are currently testing this on our engineering device where we can get 13tok/sec for KV Cache mode and 0.8tok/sec for Bert mode.
- Accuracy: We currently don't have specific accuracy metrics like perplexity available. We tested with examples such as
Prompt: "What is baseball?"
Response: "It is a game of skill, strategy, and physical ability. It is played by two teams, each with nine players. The game is played on a diamond-shaped field, with a pitcher's mound at one corner. The objective of the game is to score more runs than the opposing team by hitting the ball with a bat and running around the four bases on the field. The team with the most runs at the end of nine innings wins the game. The game of baseball is a classic and timeless game that is enjoyed by people all over the world. It is a game that is easy to learn but difficult to"

- Memory: The memory usage for both KV Cache Mode and Bert Mode are both around 11GB. We have tested on a 16GB engineering device and verified to work.

Please let me know if you have any other questions.
Thanks!

Reply:
BTW, you mentioned the memory usage for you was around 11GB. I used both "top" and "dumpsys" to check the physical RAM usage, and both showed 5GB for me on SM8650 as below. Were you using a different way to check RAM usage? I'm assuming context size is 1024 as well for your testing?

=======================================================================

Tasks: 838 total,   2 running, 836 sleeping,   0 stopped,   0 zombie
  Mem:    15267M total,    15129M used,      138M free,        1M buffers
 Swap:     9924M total,      756M used,     9167M free,     6231M cached
800%cpu  56%user   0%nice  41%sys 696%idle   0%iow   7%irq   1%sirq   0%host
  PID USER         PR  NI VIRT  RES  SHR S[%CPU] %MEM     TIME+ ARGS                                                                                                                                        
25091 shell        20   0  22G **_4.8G_** 4.7G R 72.0  32.5   1:18.94 qaihub_llama3_8b_runner --sharded_1_path qaihub_llama3_8b_token_0.pte --sharded_2_path qaihub_llama3_8b_token_1.pte --sharded_3_path qaihub+

=======================================================================

a21550@a21550:Works$ adb shell dumpsys meminfo 25091
Applications Memory Usage (in Kilobytes):
Uptime: 102398639 Realtime: 102398639
                   Pss  Private  Private     Swap      Rss     Heap     Heap     Heap
                 Total    Dirty    Clean    Dirty    Total     Size    Alloc     Free
                ------   ------   ------   ------   ------   ------   ------   ------
  Native Heap    97688    97688        0        0    97688        0        0        0
  Dalvik Heap        0        0        0        0        0        0        0        0
        Stack      392      392        0        0      392                           
    Other dev        4        0        4        0      336                           
     .so mmap     3852      356     3364        0     6720                           
   Other mmap  4984258       76  4984180        0  4984820                           
      Unknown     1456     1456        0        0     1456                           
        TOTAL  5087650    99968  4987548        0  5091412        0        0        0
 
 App Summary
                       Pss(KB)                        Rss(KB)
                        ------                         ------
           Java Heap:        0                              0
         Native Heap:    97688                          97688
                Code:     3720                           6720
               Stack:      392                            392
            Graphics:        0                              0
       Private Other:  4985716
              System:      134
             Unknown:                                 4986612
 
           TOTAL PSS:  5087650            TOTAL RSS:  5091412      TOTAL SWAP (KB):        0

=======================================================================

Reply:
Hi @a21550,
I am using top to check memory usage, and I think we are getting similar results. However, the way I am checking the memory consumption is to execute the runner and ensure no other apps are running the same time, and then I check the Mem usage, which went from:
 
 Mem:    15185M total,     5179M used,    10005M free,       14M buffers
 Swap:     6143M total,        0M used,     6143M free,     2329M cached
 
 to
 
  Mem:    15185M total,    14901M used,      283M free,        2M buffers
 Swap:     6143M total,      622M used,     5521M free,     6230M cached
 
which is around 11GB memory usage.